packages/google_sign_in/google_sign_in/lib/google_sign_in.dart

```diff
@@ -279,15 +279,15 @@ class GoogleSignIn {
   /// The optional, named parameter [canSkipCall] lets the plugin know that the
   /// method call may be skipped, if there's already [_currentUser] information.
   /// This is used from the [signIn] and [signInSilently] methods.
-  Future<GoogleSignInAccount> _addMethodCall(
+  Future<GoogleSignInAccount?> _addMethodCall(
     Function method, {
     bool canSkipCall = false,
   }) async {
-    Future<GoogleSignInAccount> response;
+    Future<GoogleSignInAccount?> response;
     if (_lastMethodCall == null) {
       response = _callMethod(method);
     } else {
-      response = _lastMethodCall.then((_) {
+      response = _lastMethodCall!.then((_) {
```

Comment:
is the analyzer able to detect the `_lastMethodCall == null` check earlier?

Reply:
I am a bit annoyed by this as well but I have seen it in many other places. My interpretation is that flow analysis only catches variables that are local to current scope. For instance:

```dart
void main() {
  String? foo;
  if (foo == null) {
    print('is null');
  } else {
    print(foo.length);
  }
}
```

is fine. However:

```dart
String? foo;

void main() {
  if (foo == null) {
    print('is null');
  } else {
    print(foo.length);
  }
}
```

is not. You need the bang operator.

Reply:
I see, thanks. I learned something new. I wonder if @munificent has any thoughts about whether this is the intended design or a temporary limitation

Reply:
It's an intended design but a known pain point that we'd like to improve if we can figure out how. The problem is that promoting fields isn't sound. In many cases, the compiler can't prove that there is not a class elsewhere that implements that class's interface and overrides with the field with a getter. Or there could be some re-entrant method call on the instance that changes the field's value between when it is checked for null and when it is later used.

Reply:
ah. that makes sense. The re-rentrant method call is interesting. Could that occur even if the method that is checking for `null` isn't async?

Reply:
Yeah. "Re-entrant" is maybe not the best word, but just consider any case between checking a field for null and accessing it where you could end up running other code on the same instance. This is contrived, but should give you the idea:

```dart
class Foo {
  int? field;

  method(Foo someFoo) {
    if (field != null) {
      someFoo.other();
      print(field + 1);
    }
  }

  other() {
    field = null;
  }
}

main() {
  var foo = Foo();
  foo.method(foo);
}
```